app/src/main/java/org/mozilla/focus/browser/VideoVoiceCommandMediaSession.kt

```diff
@@ -0,0 +1,141 @@
+                .setState(STATE_PLAYING, PLAYBACK_POSITION_UNKNOWN, 1.0f)
+                .build()
+        mediaSession.setPlaybackState(pb)
+        mediaSession.setCallback(MediaSessionCallbacks())
+        mediaSession.setFlags(MediaSessionCompat.FLAG_HANDLES_MEDIA_BUTTONS or
+                MediaSessionCompat.FLAG_HANDLES_TRANSPORT_CONTROLS)
+    }
+
+    @OnLifecycleEvent(ON_START)
+    fun onStart() {
+        mediaSession.isActive = true
+    }
+
+    @OnLifecycleEvent(ON_STOP)
+    fun onStop() {
+        // Video playback stops when backgrounded so we can end the session.
+        mediaSession.isActive = false
+    }
+
+    @OnLifecycleEvent(ON_DESTROY)
+    fun onDestroy(lifecycleOwner: LifecycleOwner) {
+        // Fragments can be re-used after onDestroy so we must remove the observer
+        // so we don't have two instances when this object is created again.
+        lifecycleOwner.lifecycle.removeObserver(this)
+
+        mediaSession.release()
+    }
+
+    fun dispatchKeyEvent(event: KeyEvent): Boolean = when (event.keyCode) {
+        // If we do nothing, these keys will be received twice: on key down by the MediaSession API
+        // and on key up by the WebView/web page, causing playback state updates to be undone.
+        // Thus we mute the WebView key up events and handle it all ourselves through MediaSession.
+        KEYCODE_MEDIA_PLAY_PAUSE, KEYCODE_MEDIA_PLAY, KEYCODE_MEDIA_PAUSE -> event.action == KeyEvent.ACTION_UP
+        else -> false
+    }
+
+    /**
+     * The methods called when a voice command is spoken; these callbacks are also called for
+     * the equivalent hardware remote button presses (e.g. "Alexa play" == hardware play/pause).
+     */
+    inner class MediaSessionCallbacks : MediaSessionCompat.Callback() {
+        override fun onPlay() {
+            onPlayPause() // See class javadoc for details.
+        }
+
+        override fun onPause() {
+            onPlayPause() // See class javadoc for details.
+        }
+
+        private fun onPlayPause() {
+            // Due to time constraints, this code is written for a single video on the page,
+            // which should cover the majority use case.
+            //
+            // We don't handle audio: see class javadoc for details.
+            iWebView?.evalJS("""
+                |(function() {
+                |    var videos = Array.from(document.querySelectorAll('video'));
+                |    if (videos.length === 0) { return; }
+                |
+                |    var videoToOperateOn = videos.find(function (video) { return !video.paused });
+                |    if (!videoToOperateOn) {
+                |        videoToOperateOn = videos[0];
+                |    }
+                |
+                |    if (videoToOperateOn.paused) {
+                |        videoToOperateOn.play();
+                |    } else {
+                |       videoToOperateOn.pause();
+                |   }
+                |})();
```

Comment:
wow nice javascript snippet - is this the intended way to do this? Or does the Amazon MediaSessions provide other APIs (rather than writing JS)?

Reply:
No other way to do this, as far as I know.

MediaSession is an Android API, not Amazon, and is intended to be an agnostic way to communicate directly with the back-end player (while MediaController communicates with the UI and MediaController communicates with MediaSession). Since the player is in the WebView, we have to communicate directly with the player and, afaik, that means JS.